sdk/tables/Azure.Data.Tables/src/DynamicTableEntity.cs

```diff
@@ -0,0 +1,221 @@
+        public Guid GetGuid(string key) => GetValue<Guid>(key);
+
+        /// <summary>
+        /// Get the value of a <see cref="DynamicTableEntity"/>'s
+        /// <see cref="Int32"/> property called
+        /// <paramref name="key"/>.
+        /// </summary>
+        /// <param name="key">The name of the property.</param>
+        /// <returns>The value of the property.</returns>
+        public int GetInt32(string key) => GetValue<int>(key);
+
+        /// <summary>
+        /// Get the value of a <see cref="DynamicTableEntity"/>'s
+        /// <see cref="Int64"/> property called
+        /// <paramref name="key"/>.
+        /// </summary>
+        /// <param name="key">The name of the property.</param>
+        /// <returns>The value of the property.</returns>
+        public long GetInt64(string key) => GetValue<long>(key);
+
+        /// <summary>
+        /// Set a document property.
+        /// </summary>
+        /// <param name="key">The property name.</param>
+        /// <param name="value">The property value.</param>
+        private protected void SetValue(string key, object value)
+        {
+            Argument.AssertNotNullOrEmpty(key, nameof(key));
+
+            if (_properties.TryGetValue(key, out object existingValue))
+            {
+                EnforceType(existingValue.GetType(), value.GetType());
+            }
+            _properties[key] = value;
+        }
+
+        /// <summary>
+        /// Get an entity property.
+        /// </summary>
+        /// <typeparam name="T">The expected type of the property value.</typeparam>
+        /// <param name="key">The property name.</param>
+        /// <returns>The value of the property.</returns>
+        private protected T GetValue<T>(string key) => (T)GetValue(key, typeof(T));
+
+        /// <summary>
+        /// Get an entity property.
+        /// </summary>
+        /// <param name="key">The property name.</param>
+        /// <param name="type">The expected type of the property value.</param>
+        /// <returns>The value of the property.</returns>
+        private protected object GetValue(string key, Type type = null)
+        {
+            if (!_properties.TryGetValue(key, out object value))
+            {
+                KeyNotFoundException exception = new KeyNotFoundException(
+                    "Could not find a member called '" + key + "' in the document.");
+                exception.Data["MissingName"] = key;
+                throw exception;
+            }
+
+            if (type != null)
+            {
+                EnforceType(type, value.GetType());
+            }
+
+            return value;
+        }
+
+        /// <summary>
+        /// Ensures that the given type matches the type of the existing
+        /// property; throws an exception if the types do not match.
+        /// </summary>
+        private static void EnforceType(Type requestedType, Type givenType)
+        {
+            if (givenType != requestedType)
```

Comment:
Should we do an `IsAssignable` check for safety?

Reply:
Do we want the user to be able to pass in an object that inherits from the accepted types? I'm not familiar with Tables usage, so I'm not sure if that would introduce any errors with serialization/deserialization.

Reply:
I meant replace the `if (givenType != requestedType)` with `if (givenType.IsAssignableFrom(requestedType))`, but I guess it's not necessary with the types you're currently supporting.  Sorry for the confusion.  (This is why you don't do PR reviews during other meetings...)

I do worry about the types if someone sets `0` which C# treats as an `int` but they intend to pass to a `long`.  Should we support coercion of numeric types?

Reply:
Type coercion raises some interesting scenarios. For example, if we are permissive here, we may let the developer make a mistake unintentionally by returning them a type they asked for even if it was not what was stored in the entity. This could potentially have unintended side effects of persisting an entity with the unintended type. Perhaps that is mostly harmless, but I imagine it could create some strange edge cases where a collection of entities end up with jagged typed properties.

Reply:
Let's file a bug and figure this out as a feature crew because anyone with a similar dictionary will have this problem.  I agree we don't want to do too much magic, but it would also be unpleasant if I initialize things to `0` and then get exceptions when I try to will it with a computed `long`.

Reply:
tracking here #13641